### PR TITLE
Fix bug in unifying records

### LINKED ISSUE #4739: Stack overflow in layout
Repro here: https://gist.github.com/joshuawarner32/4bd0d2e883c66547322765e70b065bf6

Previously discussed on zulip here: https://roc.zulipchat.com/#narrow/stream/231635-compiler-development/topic/Stack.20overflow.20in.20layout

The core of the of the repeated frames look like this:

```
    frame #355: 0x0000000102f3f124 roc`roc_mono::layout::Layout::from_var::h8282a33d2b2321dc(env=0x0000000171d75248, var=(__0 = 3000)) at layout.rs:2298:9
    frame #356: 0x0000000102f45ebc roc`roc_mono::layout::list_layout_from_elem::h4e802b6b5797e052(env=0x0000000171d75248, element_var=(__0 = 3000)) at layout.rs:4165:17
    frame #357: 0x0000000102f433fc roc`roc_mono::layout::layout_from_flat_type::h4e72dc5c77521dc7(env=0x0000000171d75248, flat_type=FlatType @ 0x00000001717622d0) at layout.rs:3050:38
    frame #358: 0x0000000102f3e438 roc`roc_mono::layout::Layout::new_help::hec2315447f591d93(env=0x0000000171d75248, _var=(__0 = 3007), content=Content @ 0x0000000171762820) at layout.rs:2237:37
    frame #359: 0x0000000102ccecac roc`roc_mono::layout::Layout::from_var::_$u7b$$u7b$closure$u7d$$u7d$::h9d3aa2ffefd225d2(env=0x0000000171d75248) at layout.rs:2300:13
    frame #360: 0x0000000102f3f484 roc`roc_mono::layout::Layout::from_var::h8282a33d2b2321dc [inlined] roc_mono::layout::Env::cached_or::h3c4da8549e816a93(self=0x0000000171d75248, var=(__0 = 3007), compute_layout={closure_env#0} @ 0x0000000171762ef8) at layout.rs:2144:9
    frame #361: 0x0000000102f3f124 roc`roc_mono::layout::Layout::from_var::h8282a33d2b2321dc(env=0x0000000171d75248, var=(__0 = 3007)) at layout.rs:2298:9
    frame #362: 0x0000000102cd61dc roc`roc_mono::layout::layout_from_recursive_union::hafc7f66777cbd3be(env=0x0000000171d75248, rec_var=(__0 = 2998), tags=0x0000000171765bf0) at layout.rs:4002:35
    frame #363: 0x0000000102f41b0c roc`roc_mono::layout::layout_from_flat_type::h4e72dc5c77521dc7(env=0x0000000171d75248, flat_type=FlatType @ 0x0000000171767520) at layout.rs:3172:13
    frame #364: 0x0000000102f3e438 roc`roc_mono::layout::Layout::new_help::hec2315447f591d93(env=0x0000000171d75248, _var=(__0 = 3010), content=Content @ 0x0000000171767a70) at layout.rs:2237:37
    frame #365: 0x0000000102ccecac roc`roc_mono::layout::Layout::from_var::_$u7b$$u7b$closure$u7d$$u7d$::h9d3aa2ffefd225d2(env=0x0000000171d75248) at layout.rs:2300:13
    frame #366: 0x0000000102f3f484 roc`roc_mono::layout::Layout::from_var::h8282a33d2b2321dc [inlined] roc_mono::layout::Env::cached_or::h3c4da8549e816a93(self=0x0000000171d75248, var=(__0 = 3010), compute_layout={closure_env#0} @ 0x0000000171768148) at layout.rs:2144:9
    frame #367: 0x0000000102f3f124 roc`roc_mono::layout::Layout::from_var::h8282a33d2b2321dc(env=0x0000000171d75248, var=(__0 = 3010)) at layout.rs:2298:9
    frame #368: 0x0000000102f42210 roc`roc_mono::layout::layout_from_flat_type::h4e72dc5c77521dc7(env=0x0000000171d75248, flat_type=FlatType @ 0x000000017176c0b0) at layout.rs:3108:51
    frame #369: 0x0000000102f3e438 roc`roc_mono::layout::Layout::new_help::hec2315447f591d93(env=0x0000000171d75248, _var=(__0 = 3000), content=Content @ 0x000000017176c600) at layout.rs:2237:37
    frame #370: 0x0000000102ccecac roc`roc_mono::layout::Layout::from_var::_$u7b$$u7b$closure$u7d$$u7d$::h9d3aa2ffefd225d2(env=0x0000000171d75248) at layout.rs:2300:13
    frame #371: 0x0000000102f3f484 roc`roc_mono::layout::Layout::from_var::h8282a33d2b2321dc [inlined] roc_mono::layout::Env::cached_or::h3c4da8549e816a93(self=0x0000000171d75248, var=(__0 = 3000), compute_layout={closure_env#0} @ 0x000000017176ccd8) at layout.rs:2144:9
```

Thus far I've had very little luck minimizing this one; it seems to be very fragile.

### PR DESCRIPTION
**Pull request by joshuawarner32**
This was leading us to have an infinitely-recursive type, which eventually causes layout to stack-overflow

Fixes #4739

### CODE REVIEW

**Comment by joshuawarner32** [comment]
Not sure why this required modifying the `glue` test? Looks to me like this potentially triggered a different bug.

**Comment by bhansconnect** [comment]
Maybe `choose_merged_var` is choosing the wrong variable and we are loosing the name? I don't actually know the code you changed, but that sounds possible.

**Comment by joshuawarner32** [comment]
Interesting - can you point me to the code that's choosing R1/R2 names? Is that in glue, or somewhere deeper in the compiler?

**Comment by joshuawarner32** [comment]
Ok, I think I figured it out - I accidentally reversed some of the arguments there, which meant `choose_merged_var` choose a different default value if neither type were a `RecursiveVar`.

File: crates/compiler/solve/tests/solve_expr.rs [hunk 1]
```diff
@@ -522,6 +522,41 @@ mod solve_expr {
         );
     }
 
+    #[test]
+    fn my_bug() {
```
Comment (ayazhafiz) [inline crates/compiler/solve/tests/solve_expr.rs:526]: Can this test be given a more descriptive name, e.g. "choose_correct_recursion_var_under_record"?

File: crates/compiler/solve/tests/solve_expr.rs [hunk 2]
```diff
@@ -522,6 +522,41 @@ mod solve_expr {
...
+        Record fields ->
+            fields
+                |> List.map \f ->
+                    printed = printCombinatorParser f.parser
+                    if Bool.false then printed else "foo"
+                |> List.first
+                |> Result.withDefault ("foo")
+
+printCombinatorParser (Record [])
+    "#
```
Comment (ayazhafiz) [inline crates/compiler/solve/tests/solve_expr.rs:551]:
```suggestion
                r#"
                Parser : [
                    Specialize Parser,
                    Record (List {parser: Parser}),
                ]
                
                printCombinatorParser : Parser -> Str
                printCombinatorParser = \parser ->
                    when parser is
                #        ^^^^^^
                        Specialize p ->
                            printed = printCombinatorParser p
                            if Bool.false then printed else "foo"
                        Record fields ->
                            fields
                                |> List.map \f ->
                                    printed = printCombinatorParser f.parser
                                    if Bool.false then printed else "foo"
                                |> List.first
                                |> Result.withDefault ("foo")
                
                printCombinatorParser (Record [])
                "#
```

to line up with how other tests of this type are indented in this file

**Review by ayazhafiz — COMMENTED** [review]
Congratulations for resolution through the trials and tribulations of the type system!!

**Review by ayazhafiz — CHANGES_REQUESTED** [review]
Let's also add a test to `test_mono` that ensures that this actually passes layout generation.

**Comment by joshuawarner32** [comment]
Unfortunately test_mono isn't happy ☹️ 

```
thread 'choose_correct_recursion_var_under_record' panicked at 'IR problems found:
── SYMBOL LAYOUT DOESN'T MATCH ITS USE ─────────────────────────────────────────

in main : () -> Str (no niche)

0│  procedure `#UserApp.main` ():
1│>     let `#UserApp.28` : List [<r>C List *self, C *self] = Array [];
2│      let `#UserApp.15` : [<r>C List *self, C *self] = TagId(0) `#UserApp.28`;

#UserApp.28 defined here with layout List [<r>C List *self, C *self]

1│      let `#UserApp.28` : List [<r>C List *self, C *self] = Array [];
2│>     let `#UserApp.15` : [<r>C List *self, C *self] = TagId(0) `#UserApp.28`;
3│      let `#UserApp.14` : Str = CallByName `#UserApp.printCombinatorParser` `#UserApp.15`;

#UserApp.28 used as a tag's payload here with layout List *self
── SYMBOL INITIALIZER HAS THE WRONG LAYOUT ─────────────────────────────────────

in printCombinatorParser : ([<r>C List *self, C *self]) -> Str (no niche)

8 │          ret `#UserApp.17`;
9 │>     else
10│          let `#UserApp.fields` : List [<r>C List *self, C *self] = UnionAtIndex (Id 0) (Index 0) `#UserApp.parser`;

#UserApp.fields is defined as List [<r>C List *self, C *self] but its
initializer is List *self', crates/compiler/test_mono/src/tests.rs:173:5
```

Here's the full ir:

```
procedure `#UserApp.main` ():
    let `#UserApp.28` : List [<r>C List *self, C *self] = Array [];
    let `#UserApp.15` : [<r>C List *self, C *self] = TagId(0) `#UserApp.28`;
    let `#UserApp.14` : Str = CallByName `#UserApp.printCombinatorParser` `#UserApp.15`;
    dec `#UserApp.15`;
    ret `#UserApp.14`;
procedure `#UserApp.printCombinatorParser` (`#UserApp.parser`):
    let `#UserApp.25` : U8 = 1i64;
    let `#UserApp.26` : U8 = GetTagId `#UserApp.parser`;
    let `#UserApp.27` : Int1 = lowlevel Eq `#UserApp.25` `#UserApp.26`;
    if `#UserApp.27` then
        let `#UserApp.p` : [<r>C List *self, C *self] = UnionAtIndex (Id 1) (Index 0) `#UserApp.parser`;
        let `#UserApp.printed` : Str = CallByName `#UserApp.printCombinatorParser` `#UserApp.p`;
        let `#UserApp.17` : Str = "foo";
        ret `#UserApp.17`;
    else
        let `#UserApp.fields` : List [<r>C List *self, C *self] = UnionAtIndex (Id 0) (Index 0) `#UserApp.parser`;
        let `#UserApp.22` : {} = Struct {};
        let `#UserApp.21` : List Str = CallByName `List.map` `#UserApp.fields` `#UserApp.22`;
        let `#UserApp.19` : [C {}, C Str] = CallByName `List.first` `#UserApp.21`;
        dec `#UserApp.21`;
        let `#UserApp.20` : Str = "foo";
        let `#UserApp.18` : Str = CallByName `Result.withDefault` `#UserApp.19` `#UserApp.20`;
        dec `#UserApp.20`;
        ret `#UserApp.18`;
```

**Comment by ayazhafiz** [comment]
That’s okay, you can mark the proc macro as #[test_mono(no_check)]. I plan to fix that false negative when i’m back from vacation
________________________________

**Comment by ayazhafiz** [comment]
(or, if you’re interested in another fun dive, feel free to fix that false positive too - should be a short project)
________________________________

**Review by ayazhafiz — APPROVED** [review]

**Pull request merged**